sdk/Pulumi/Provider/Provider.cs

```diff
@@ -363,13 +363,15 @@ public DeleteRequest(Urn urn, string id, ImmutableDictionary<string, PropertyVal
 
     public sealed class ConstructRequest
     {
+        public ProviderResource? Provider { get; init; }
```

Comment:
Arguably this should be non-nullable and throw with a message like "self provider references aren't supported update your pulumi CLI", that would match how we handle other cases like this that depend on engine support.